File: iosApp/iosApp/Modules/Dashboard/DashboardWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultDashboardWireframe {

    private weak var parent: UIViewController?
    private let accountWireframeFactory: AccountWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let mnemonicConfirmationWireframeFactory: MnemonicConfirmationWireframeFactory
    private let currencyPickerWireframeFactory: CurrencyPickerWireframeFactory
    private let currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory
    private let currencySendWireframeFactory: CurrencyCurrencyWireframeFactory
    private let currencySwapWireframeFactory: CurrencySwapWireframeFactory
    private let nftDetailWireframeFactory: NFTDetailWireframeFactory
    private let qrCodeScanWireframeFactory: QRCodeScanWireframeFactory
    private let themePickerWireframeFactory: ThemePickerWireframeFactory
    private let improvementProposalsWireframeFactory: ImprovementProposalsWireframeFactory
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    private let walletService: WalletService
    private let nftsService: NFTsService
    private let actionsService: ActionsService

    private weak var vc: UIViewController?

    init(
        _ parent: UIViewController?,
        accountWireframeFactory: AccountWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        mnemonicConfirmationWireframeFactory: MnemonicConfirmationWireframeFactory,
        currencyPickerWireframeFactory: CurrencyPickerWireframeFactory,
        currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory,
        currencySendWireframeFactory: CurrencyCurrencyWireframeFactory,
        currencySwapWireframeFactory: CurrencySwapWireframeFactory,
        nftDetailWireframeFactory: NFTDetailWireframeFactory,
        qrCodeScanWireframeFactory: QRCodeScanWireframeFactory,
        themePickerWireframeFactory: ThemePickerWireframeFactory,
        improvementProposalsWireframeFactory: ImprovementProposalsWireframeFactory,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService,
        walletService: WalletService,
        nftsService: NFTsService,
        actionsService: ActionsService
    ) {
        self.parent = parent
        self.accountWireframeFactory = accountWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.mnemonicConfirmationWireframeFactory = mnemonicConfirmationWireframeFactory
        self.currencyPickerWireframeFactory = currencyPickerWireframeFactory
        self.currencyReceiveWireframeFactory = currencyReceiveWireframeFactory
        self.currencySendWireframeFactory = currencySendWireframeFactory
        self.currencySwapWireframeFactory = currencySwapWireframeFactory
        self.nftDetailWireframeFactory = nftDetailWireframeFactory
        self.qrCodeScanWireframeFactory = qrCodeScanWireframeFactory
        self.themePickerWireframeFactory = themePickerWireframeFactory
        self.improvementProposalsWireframeFactory = improvementProposalsWireframeFactory
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
        self.walletService = walletService
        self.nftsService = nftsService
        self.actionsService = actionsService
    }
}

extension DefaultDashboardWireframe {

    func present() {
        let vc = wireUp()
        if let tabVc = parent as? UITabBarController {
            let vcs = tabVc.viewControllers ?? [] + [vc]
            tabVc.setViewControllers(vcs, animated: false)
        } else {
            parent?.show(vc, sender: self)
        }
    }

    func navigate(with destination: DashboardWireframeDestination) {
        guard let vc = self.vc ?? parent else {
            print("DefaultDashboardWireframe has no view")
            return
        }
        if destination is DashboardWireframeDestination.KeyStoreNetworkSettings {
            vc.edgeCardsController?.setDisplayMode(.overview, animated: true)
        }
        if destination is DashboardWireframeDestination.PresentUnderConstructionAlert {
            alertWireframeFactory.make(parent, context: .underConstructionAlert()).present()
        }
        if destination is DashboardWireframeDestination.ScanQRCode {
            guard let network = networksService.network else { return }
            let context = QRCodeScanWireframeContext(
                type: QRCodeScanWireframeContext.Type_Network(network: network),
                handler: navigateToCurrencySend()
            )
            qrCodeScanWireframeFactory.make(vc, context: context).present()
        }
        if destination is DashboardWireframeDestination.Receive {
            navigateToReceive()
        }
        if let input = destination as? DashboardWireframeDestination.Send {
            guard let network = networksService.network, let address = input.addressTo else {
                navigateToCurrencyPickerAndSend()
                return
            }
            currencySendWireframeFactory.make(
                vc,
                context: .init(network: network, address: address, currency: nil)
            ).present()
        }
        if destination is DashboardWireframeDestination.Swap {
            guard let network = networksService.network else { return }
            currencySwapWireframeFactory.make(
                vc,
                context: .init(network: network, currencyFrom: nil, currencyTo: nil)
            ).present()
        }
        if destination is DashboardWireframeDestination.MnemonicConfirmation {
            mnemonicConfirmationWireframeFactory.make(parent).present()
        }
        if destination is DashboardWireframeDestination.ThemePicker {
            themePickerWireframeFactory.make(vc).present()
        }
        if destination is DashboardWireframeDestination.ImprovementProposals {
            improvementProposalsWireframeFactory.make(vc).present()
        }
        if let input = destination as? DashboardWireframeDestination.EditCurrencies {
            let onCompletion: (([CurrencyPickerWireframeContext.Result]) -> Void) = { result in
                guard let currencies = result.first?.selectedCurrencies else { return }
                input.onCompletion(currencies)
            }
            currencyPickerWireframeFactory.make(
                vc,
                context: .init(
                    isMultiSelect: true,
                    showAddCustomCurrency: true,
                    networksData: [
                        .init(network: input.network, favouriteCurrencies: input.selectedCurrencies, currencies: nil)
                    ],
                    selectedNetwork: nil,
                    handler: onCompletion
                )
            ).present()
        }
        if let input = destination as? DashboardWireframeDestination.Wallet {
            accountWireframeFactory.make(
                vc,
                context: .init(network: input.network, currency: input.currency)
            ).present()
        }
        if let input = destination as? DashboardWireframeDestination.NftItem {
            guard let vc = self.vc else { return }
            let context = NFTDetailWireframeContext(
                nftId: input.nft.identifier,
                collectionId: input.nft.collectionIdentifier
            )
            nftDetailWireframeFactory.make(vc, context: context).present()
        }
    }
}

private extension DefaultDashboardWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultDashboardInteractor(
            networksService: networksService,
            currencyStoreService: currencyStoreService,
            walletService: walletService,
            nftsService: nftsService,
            actionsService: actionsService
        )
        let vc: DashboardViewController = UIStoryboard(.dashboard).instantiate()
        let presenter = DefaultDashboardPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        let nc = NavigationController(rootViewController: vc)
        self.vc = nc
        return nc
    }
    
    func navigateToCurrencyPickerAndSend() {
        guard let vc = self.vc else { return }
        let onCompletion: (([CurrencyPickerWireframeContext.Result]) -> Void) = { [weak self] result in
            guard
                let self = self,
                let network = result.first?.network,
                let currency = result.first?.selectedCurrencies.first
            else { return }
            self.currencySendWireframeFactory.make(
                vc.presentedViewController ?? vc,
                context: .init(network: network, address: nil, currency: currency)
            ).present()
        }
        currencyPickerWireframeFactory.make(
            vc,
            context: .init(
                isMultiSelect: false,
                showAddCustomCurrency: false,
                networksData: networksService.enabledNetworks().compactMap{
                    .init(network: $0, favouriteCurrencies: nil, currencies: nil)
                },
                selectedNetwork: nil,
                handler: onCompletion
            )
        ).present()
    }
    
    func navigateToCurrencySend() -> (String) -> Void {
        { [weak self] addressTo in self?.navigate(with: DashboardWireframeDestination.Send(addressTo: addressTo)) }
    }

    func navigateToReceive() {
        let onCompletion: (([CurrencyPickerWireframeContext.Result]) -> Void) = {
            [weak self] result in
            guard let self = self else { return }
            guard
                let network = result.first?.network,
                let currency = result.first?.selectedCurrencies.first
            else { return }
            self.currencyReceiveWireframeFactory.make(
                self.vc?.presentedViewController ?? self.vc,
                context: .init(network: network, currency: currency)
            ).present()
        }
        currencyPickerWireframeFactory.make(
            vc,
            context: .init(
                isMultiSelect: false,
                showAddCustomCurrency: false,
                networksData: networksService.enabledNetworks().compactMap{
                    .init(network: $0, favouriteCurrencies: nil, currencies: nil)
                },
                selectedNetwork: nil,
                handler: onCompletion
            )
        ).present()
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/dashboard/DashboardWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.dashboard

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3wallet.android.modules.compose.account.AccountWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencypicker.CurrencyPickerWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyreceive.CurrencyReceiveWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencysend.CurrencySendWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyswap.CurrencySwapWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.improvementproposals.ImprovementProposalsWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicconfirmation.MnemonicConfirmationWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.nftdetail.NFTDetailWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.qrcodescan.QRCodeScanWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.account.AccountWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencyPicker.CurrencyPickerWireframeContext.NetworkData
import com.sonsofcrypto.web3walletcore.modules.currencyReceive.CurrencyReceiveWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencySend.CurrencySendWireframeContext
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapWireframeContext
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframe
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.EditCurrencies
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.ImprovementProposals
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.KeyStoreNetworkSettings
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.MnemonicConfirmation
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.NftItem
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.Receive
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.ScanQRCode
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.Send
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.Swap
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.ThemePicker
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframeDestination.Wallet
import com.sonsofcrypto.web3walletcore.modules.dashboard.DefaultDashboardInteractor
import com.sonsofcrypto.web3walletcore.modules.dashboard.DefaultDashboardPresenter
import com.sonsofcrypto.web3walletcore.modules.nftDetail.NFTDetailWireframeContext
import com.sonsofcrypto.web3walletcore.modules.qrCodeScan.QRCodeScanWireframeContext
import com.sonsofcrypto.web3walletcore.services.actions.ActionsService
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService

class DefaultDashboardWireframe(
    private val parent: WeakRef<Fragment>?,
    private val networksService: NetworksService,
    private val currencyStoreService: CurrencyStoreService,
    private val walletService: WalletService,
    private val nftsService: NFTsService,
    private val actionsService: ActionsService,
    private val currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory,
    private val currencySendWireframeFactory: CurrencySendWireframeFactory,
    private val currencySwapWireframeFactory: CurrencySwapWireframeFactory,
    private val currencyPickerWireframeFactory: CurrencyPickerWireframeFactory,
    private val accountWireframeFactory: AccountWireframeFactory,
    private val nftDetailWireframeFactory: NFTDetailWireframeFactory,
    private val mnemonicConfirmationWireframeFactory: MnemonicConfirmationWireframeFactory,
    private val improvementProposalsWireframeFactory: ImprovementProposalsWireframeFactory,
    private val qrCodeScanWireframeFactory: QRCodeScanWireframeFactory,
): DashboardWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.push(fragment, animated = true)
    }

    override fun navigate(destination: DashboardWireframeDestination) {
        when (destination) {
            is Wallet -> {
                val context = AccountWireframeContext(destination.network, destination.currency)
                accountWireframeFactory.make(parent?.get(), context).present()
            }
            is KeyStoreNetworkSettings -> {
                println("[AA] navigate to KeyStoreNetworkSettings")
            }
            is ScanQRCode -> {
                val network = networksService.network ?: return
                val context = QRCodeScanWireframeContext(
                    type = QRCodeScanWireframeContext.Type.Network(network),
                    handler = { addressTo -> navigate(Send(addressTo)) }
                )
                qrCodeScanWireframeFactory.make(parent?.get(), context).present()
            }
            is MnemonicConfirmation -> {
                mnemonicConfirmationWireframeFactory.make(parent?.get()).present()
            }
            is ThemePicker -> {
                println("[AA] navigate to ThemePicker")
            }
            is ImprovementProposals -> {
                improvementProposalsWireframeFactory.make(parent?.get()).present()
            }
            is Receive -> {
                navigateToReceive()
            }
            is Send -> {
                navigateToSend(destination)
            }
            is Swap -> {
                val network = networksService.network ?: return
                val context = CurrencySwapWireframeContext(network, null, null)
                currencySwapWireframeFactory.make(parent?.get(), context).present()
            }
            is EditCurrencies -> {
                navigateToEditCurrencies(destination)
            }
            is NftItem -> {
                val context = NFTDetailWireframeContext(
                    destination.nft.identifier, destination.nft.collectionIdentifier
                )
                nftDetailWireframeFactory.make(parent?.get(), context).present()
            }
            else -> { println("[AA] navigate to $destination") }
        }
    }

    private fun wireUp(): Fragment {
        val view = DashboardFragment()
        val interactor = DefaultDashboardInteractor(
            networksService,
            currencyStoreService,
            walletService,
            nftsService,
            actionsService,
        )
        val presenter = DefaultDashboardPresenter(
            WeakRef(view),
            this,
            interactor
        )
        view.presenter = presenter
        return view
    }

    private fun navigateToReceive() {
        val context = CurrencyPickerWireframeContext(
            isMultiSelect = false,
            showAddCustomCurrency = false,
            networksData = networksService.enabledNetworks().map {
                NetworkData(it, null, null)
            },
            selectedNetwork = null,
            handler = {
                it.firstOrNull()?.let { result ->
                    result.selectedCurrencies.firstOrNull()?.let { currency ->
                        val context = CurrencyReceiveWireframeContext(
                            network = result.network,
                            currency= currency,
                        )
                        currencyReceiveWireframeFactory.make(
                            parent?.get(), context
                        ).present()
                    }
                }
            }
        )
        currencyPickerWireframeFactory.make(parent?.get(), context).present()
    }

    private fun navigateToSend(destination: Send) {
        if (destination.addressTo != null) {
            val network = networksService.network ?: return
            val context = CurrencySendWireframeContext(
                network = network,
                address = destination.addressTo,
                currency= null,
            )
            currencySendWireframeFactory.make(parent?.get(), context).present()
        } else {
            val context = CurrencyPickerWireframeContext(
                isMultiSelect = false,
                showAddCustomCurrency = false,
                networksData = networksService.enabledNetworks().map {
                    NetworkData(it, null, null)
                },
                selectedNetwork = null,
                handler = {
                    it.firstOrNull()?.let { result ->
                        result.selectedCurrencies.firstOrNull()?.let { currency ->
                            val context = CurrencySendWireframeContext(
                                network = result.network,
                                address = null,
                                currency= currency,
                            )
                            currencySendWireframeFactory.make(parent?.get(), context).present()
                        }
                    }
                }
            )
            currencyPickerWireframeFactory.make(parent?.get(), context).present()
        }
    }

    private fun navigateToEditCurrencies(destination: EditCurrencies) {

        val context = CurrencyPickerWireframeContext(
            isMultiSelect = true,
            showAddCustomCurrency = true,
            networksData = listOf(
                NetworkData(destination.network, destination.selectedCurrencies, null)
            ),
            selectedNetwork = null,
            handler = {
                if (it.firstOrNull() != null) {
                    destination.onCompletion(it.first().selectedCurrencies)
                }
            }
        )
        currencyPickerWireframeFactory.make(parent?.get(), context).present()
    }
}